### PR TITLE
Added check for torch.where on CPU that both arguments have same dtype

### PR DESCRIPTION
**Pull request by nairbv**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30662 Added check for torch.where on CPU that both arguments have same dtype**
* #30661 Make zeros argument of torch.where same dtype as other argument

Cherry picked from: https://github.com/pytorch/pytorch/pull/29081

Co-authored-by: ifedan

Differential Revision: [D18782295](https://our.internmc.facebook.com/intern/diff/D18782295)

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -763,6 +763,45 @@ def test_where_bool_tensor(self):
             res = torch.where(a > 0)
             self.assertEqual(1, len(res))
 
+    def test_where_tensor(self):
+        def rand_tensor(size, dtype, device):
```
Comment (zou3519) [inline test/test_torch.py]: Could we please deduplicate the rand_tensor calls? This is the third place we have logic to create a random tensor; this looks pretty similar to the one in test_type_promotion.

Comment (nairbv) [reply test/test_torch.py]: as in the other review, there's plenty we could do, but I want master to match 1.3.1 first.

Comment (nairbv) [reply test/test_torch.py]: this is just a cherry pick from 1.3.1 to master, also need to merge to 1.4

Comment (zou3519) [reply test/test_torch.py]: I'm pretty uncomfortable with having the same code duplicated. Could you please send a followup PR to deduplicate this and https://github.com/pytorch/pytorch/blob/8b29701ae505a17595a85cac05109ddf3f82e68e/test/test_type_promotion.py#L174-L187 ?

**Review by zou3519 — APPROVED** [review]

File: test/test_torch.py (hunk 1 above)
Comment (nairbv) [reply test/test_torch.py]: there's an issue filed. I added a reference to this in the issue tracker:
https://github.com/pytorch/pytorch/issues/30611

**Review by gchanan — APPROVED** [review]

**Pull request closed without merging**